cpp/src/arrow/record_batch.cc

```diff
@@ -37,6 +37,26 @@
 
 namespace arrow {
 
+namespace {
+// If there will only be one slice returned it is cheaper to just return the original
+// RecordBatch (no overhead from vector and std::shared_ptr copying on underlying arrays).
+
+struct SliceIteratorFunctor {
+  Result<std::shared_ptr<RecordBatch>> Next() {
+    if (current_offset < batch->num_rows()) {
+      std::shared_ptr<RecordBatch> next = batch->Slice(current_offset, slice_size);
+      current_offset += slice_size;
+      return next;
+    }
+    return IterationTraits<std::shared_ptr<RecordBatch>>::End();
+  }
+  const RecordBatch* const batch;
```

Comment:
yes, it would.  The tricky part is we can't get a shared_ptr to record batch if this is a member method.  (unless we add enable_shared_from_this).

To avoid any contention here for now I've moved this to be an implementation detail.